Validate vlan mtu not greater than base iface

adwait-thattey opened the pull request:
fixes #1047 
A new validator is added that verifies that the
MTU of VLAN is less than or equal to MTU of base iface

Signed-off-by: Adwait Thattey <coderdude1999@gmail.com>

adwait-thattey commented:
@ffmancera Can u please check if this is what is required.
If yes, I will make changes to the tests (if required)

ffmancera commented on libnmstate/validator.py:
```diff
@@ -400,6 +402,18 @@ def _validate_linux_bond(original_desired_state, current_state):
+        if iface_state[Interface.TYPE] == InterfaceType.VLAN:
+            base_iface_name = iface_state.get(VLAN.CONFIG_SUBTREE).get(VLAN.BASE_IFACE)
+            base_iface_cur_state = current_state.interfaces.get(base_iface_name)
+            if base_iface_cur_state[Interface.TYPE] == InterfaceType.ETHERNET:
```
IMO, this check is not required. You can remove it, I think.

ffmancera commented on libnmstate/validator.py:
```diff
@@ -400,6 +402,18 @@ def _validate_linux_bond(original_desired_state, current_state):
+                base_iface_mtu = base_iface_cur_state.get(Interface.MTU)
+                vlan_mtu = iface_state.get(Interface.MTU)
+                if vlan_mtu > base_iface_mtu:
+                    raise ValueError("MTU of VLAN can not be more than MTU of base Interface")
```
It would be nice if the error shows up the information, something like: "vlan1 mtu (number) cannot be greater than its base interface eth1 mtu (number)". It should be easy because you already have all the information stored in variables.

adwait-thattey commented:
Done. I am running tests locally. Will push after that

One doubt:
Is it okay to directly put the validation in `_validate_vlan` or should I create another method
`_assert_vlan_mtu_not_greater_than_base_iface` and call that from the `_validate_vlan`

because I find the latter approach used at multiple places in `validator.py`

adwait-thattey commented:
I had to make a few changes considering a few existing tests

1. Added condition to check only if VLAN is up. 
2. Changed an existing test that was setting MTU greater than the default value of base eth

After this, some tests are still failing. In these tests, the base and the VLAN are being created at the same time. So the base exist in the desired_state and not the current_state and the error is thrown.
How do you propose I handle this? 

Should I check for the base first in the desired state and then in the current state?
In such a situation, if the value of base MTU is not provided in desired_state, should I default it to 1500?

adwait-thattey commented:
squashed the commits

Also, I am thinking of adding a test that deliberately adds a VLAN with MTU greater than the base to check if an error is thrown

ffmancera commented:
> squashed the commits
> 
> Also, should I add a test that deliberately adds a VLAN with MTU greater than the base to check if an error is thrown?

Yes, you should do it.

adwait-thattey commented:
Updated with validator
I tried writing the test for this like this in [vlan_test.py](https://github.com/nmstate/nmstate/blob/master/tests/integration/vlan_test.py)

```python
def test_add_vlan_with_mtu_greater_than_base(eth1_up):
    desired_state = {
        Interface.KEY: [
            {
                Interface.NAME: VLAN_IFNAME,
                Interface.TYPE: InterfaceType.VLAN,
                Interface.STATE: InterfaceState.UP,
                VLAN.TYPE: {VLAN.ID: 101, VLAN.BASE_IFACE: eth1_up[Interface.KEY][0][Interface.NAME]},
                Interface.MTU: eth1_up[Interface.MTU] + 1
            }
        ]
    }

    with pytest.raises(ValueError):
        libnmstate.apply(desired_state)
```
The problem I am facing is that `eth1_up` does not seem to have a parameter "mtu". Thus it gives a KeyError. How do I get the mtu of eth1?

adwait-thattey commented:
Added test and Rebased

ffmancera commented:
Looking into this tomorrow. Thanks for your contribution! :-)

ffmancera commented on tests/integration/ethernet_mtu_test.py:
```diff
@@ -158,7 +158,7 @@ def test_set_mtu_on_two_vlans_with_a_shared_base(eth1_up):
         }
         for iface_state in desired_state[Interface.KEY]:
-            iface_state[Interface.MTU] = 2000
+            iface_state[Interface.MTU] = 1500
```
Why are you modifying this test?

ffmancera commented on libnmstate/validator.py.bak:
```diff
@@ -0,0 +1,498 @@
+        logging.warning(
+            "Interfaces count exceeds the limit %s in desired state",
+            MAX_SUPPORTED_INTERFACES,
+        )
```
I think this file is wrong.

ffmancera commented on libnmstate/validator.py:
```diff
@@ -357,6 +359,40 @@ def _assert_vlan_filtering_trunk_tag(trunk_tag_state):
+                    base_iface_state = current_state.interfaces.get(
+                        base_iface_name
+                    )
+                if not base_iface_state:
+                    raise ValueError(
+                        f"base interface {base_iface_name} for vlan \
+                        {iface_name} does not exist"
+                    )
```
This if statement could be inside of the one above, this way it improves readability.

ffmancera commented on libnmstate/validator.py:
```diff
@@ -400,6 +436,12 @@ def _validate_linux_bond(original_desired_state, current_state):
             _assert_bond_config_is_legal(iface_state, original_iface_state)
 
 
+def _validate_vlan(original_desired_state, current_state):
+    _assert_vlan_mtu_not_greater_than_base_iface(
+        original_desired_state, current_state
+    )
+
+
```
In this case the function only contains another function. There is no need for this alias, IMO.

ffmancera commented on libnmstate/validator.py:
```diff
@@ -357,6 +359,40 @@ def _assert_vlan_filtering_trunk_tag(trunk_tag_state):
+
+                base_iface_mtu = base_iface_state.get(Interface.MTU)
+                vlan_mtu = iface_state.get(Interface.MTU)
+                if vlan_mtu and base_iface_mtu and (vlan_mtu > base_iface_mtu):
```
I think `vlan_mtu` and `base_iface_mtu` cannot be `None` at this point. I suggest you to get them directly with `[]` and there is no need to check them here. Thanks!

adwait-thattey replied:
In one (or maybe more) test cases they are coming as None
`test_create_vlan_over_a_bond_slave` is one of the tests failing if I don't check that.

Full past failure log: https://pastebin.com/Jf4nVPNB 
from https://travis-ci.com/github/nmstate/nmstate/jobs/336141093

ffmancera replied:
Fair enough, thanks!

ffmancera commented on libnmstate/validator.py:
```diff
@@ -357,6 +359,40 @@ def _assert_vlan_filtering_trunk_tag(trunk_tag_state):
+def _assert_vlan_mtu_not_greater_than_base_iface(
+    original_desired_state, current_state
+):
+    # validate if vlan mtu is greater than base interface mtu
+    for iface_name, iface_state in original_desired_state.interfaces.items():
+        if iface_state.get(Interface.TYPE) == InterfaceType.VLAN:
+            if iface_state.get(Interface.STATE) == InterfaceState.UP:
+                base_iface_name = iface_state.get(VLAN.CONFIG_SUBTREE).get(
+                    VLAN.BASE_IFACE
+                )
+                # check if base_iface is in desired state.
+                # if not then get from current state
```
Please, remove the comments.

adwait-thattey commented:
Made the changes

ffmancera commented:
Travis CI seems blocked, I am restarting it..

ffmancera commented:
For some reason, CI is not working on this PR. All the jobs are passing but it is not reporting back. Do you why? @tyll @cathay4t?

tyll commented:
> For some reason, CI is not working on this PR. All the jobs are passing but it is not reporting back. Do you know why? @tyll @cathay4t?

I guess it is a Travis problem, saw this with the network role, too. Best would be to rebase which needs to be done anyhow IMHO.

adwait-thattey commented:
@ffmancera  Rebased
But one of the builds is failing with 
```docker exec -t -i f4165a6c906e2d2bb56edca298ca15b689d44c45a762bd685d1517be538f9897 /bin/bash -c 'cd /workspace/nmstate && dnf clean all'
error: db5 error(17) from dbenv->open: File exists
error: cannot open Packages index using db5 - File exists (17)
error: cannot open Packages database in /var/lib/rpm
Error: Error: rpmdb open failed
```
Not sure what this is due to...

ffmancera commented:
We have made a lot of changes in the validator, could you rebase this an adapt it to the new code? The code looks good but it needs to be adapted to the new design. Thanks!

adwait-thattey commented:
@ffmancera rebased and adapted the code for the new design
Please review

cathay4t commented on libnmstate/ifaces/ifaces.py:
```diff
@@ -165,6 +166,35 @@ def _validate_ovs_patch_peers(self):
 
+    def _validate_vlan_mtu_lt_base_mtu(self):
+        for iface in self._ifaces.values():
+            if iface.iface_type == InterfaceType.VLAN:
```
Should also handle `Vxlan`.

cathay4t commented on libnmstate/ifaces/ifaces.py:
```diff
@@ -165,6 +166,35 @@ def _validate_ovs_patch_peers(self):
                         )
 
+    def _validate_vlan_mtu_lt_base_mtu(self):
+        for iface in self._ifaces.values():
```
Should check whether `iface.mtu` is valid first.

adwait-thattey replied:
Should I raise NmStateValueError if mtu of vlan is not present? Or just skip the check then

cathay4t commented on libnmstate/ifaces/ifaces.py:
```diff
@@ -165,6 +166,35 @@ def _validate_ovs_patch_peers(self):
+                    vlan_mtu = iface.mtu
+                    if (
+                        vlan_mtu
+                        and base_iface_mtu
```
When base_iface/parent interface does not have MTU defined while VLAN/VxLAN does, we should set parent interface MTU as vlan's.
But it could be done in another PR if you don't want to do it.

adwait-thattey replied:
I have added this in the validator itself. 
is that fine or should it be added somewhere else?

cathay4t commented on tests/integration/vlan_test.py:
```diff
@@ -192,6 +192,26 @@ def test_add_new_base_iface_with_vlan():
         libnmstate.apply(desired_state, verify_change=False)
 
 
+def test_add_vlan_with_mtu_greater_than_base(eth1_up):
```
Also need unit test case to cover all the cases.

adwait-thattey commented:
@cathay4t 
If mtu for vlan doesn't exist, I am raising an exception. [Ref. ifaces/ifaces.py, Line 184](https://github.com/adwait-thattey/nmstate/blob/fix-issue-1047/libnmstate/ifaces/ifaces.py#L184)

This is causing some [other tests](https://travis-ci.com/github/nmstate/nmstate/jobs/351085970#L1246) to fail 

Instead of raising an exception should I just skip the validator if mtu for vlan doesn't exist?

cathay4t commented on libnmstate/ifaces/ifaces.py:
```diff
@@ -165,6 +166,43 @@ def _validate_ovs_patch_peers(self):
+                            f"base interface {iface.parent} for vlan \
+                                            {iface.name} does not exist"
+                        )
+
```
This is my second time to say it:


This is not required.
The `Ifaces._validate_unknown_parent()` already checked it. No need to do it again.

cathay4t commented on libnmstate/ifaces/ifaces.py:
```diff
@@ -165,6 +166,43 @@ def _validate_ovs_patch_peers(self):
+
+                    if iface.mtu > base_iface.mtu:
+                        raise NmstateValueError(
+                            f"{iface.name} mtu {iface.mtu} cannot be \
```
The use of `\` is incorrect: by using it, you add a lot spaces.

You should use:

```python
raise NmstateValueError(
    f"Interface {iface.name} has bigger "
    f"MTU({iface.mtu}) "
    f"than its base interface: {iface.parent} "
    f"MTU({parent_iface.mtu})"
)
```

cathay4t commented on libnmstate/ifaces/ifaces.py:
```diff
@@ -165,6 +166,43 @@ def _validate_ovs_patch_peers(self):
+                    if not iface.mtu:
+                        raise NmstateValueError(
+                            f"paramater mtu missing for vlan {iface.name}"
+                        )
```
Should not raise exception as MTU is OK to not defined.

cathay4t commented on libnmstate/ifaces/ifaces.py:
```diff
@@ -165,6 +166,43 @@ def _validate_ovs_patch_peers(self):
+                or iface.iface_type == InterfaceType.VXLAN
+            ):
+
+                if iface.state == InterfaceState.UP:
```
Use `iface.is_up`.

cathay4t commented:
> @cathay4t
> If mtu for vlan doesn't exist, I am raising an exception. [Ref. ifaces/ifaces.py, Line 184](https://github.com/adwait-thattey/nmstate/blob/fix-issue-1047/libnmstate/ifaces/ifaces.py#L184)
> 
> This is causing some [other tests](https://travis-ci.com/github/nmstate/nmstate/jobs/351085970#L1246) to fail
> 
> Instead of raising an exception should I just skip the validator if mtu for vlan doesn't exist?

I will look into the failures once you fixed all my changes requests.

adwait-thattey commented:
@cathay4t made the changes. Tests are also passing

cathay4t commented on libnmstate/ifaces/ifaces.py:
```diff
@@ -165,6 +166,32 @@ def _validate_ovs_patch_peers(self):
                         )
 
+    def _validate_vlan_mtu_lte_base_mtu(self):
+        """
+            Validate that mtu of vlan or vxlan is less than
+            or equal to it's base interface's MTU
+
+            If base MTU is not present, set same as vlan MTU
```
Not indent correctly.

cathay4t commented on libnmstate/ifaces/ifaces.py:
```diff
@@ -165,6 +166,32 @@ def _validate_ovs_patch_peers(self):
                             " patch port"
                         )
 
+    def _validate_vlan_mtu_lte_base_mtu(self):
```
lte is not a common abbreviation. Use `_validate_vlan_mtu` is enough.

cathay4t commented on libnmstate/ifaces/ifaces.py:
```diff
@@ -165,6 +166,32 @@ def _validate_ovs_patch_peers(self):
+            If base MTU is not present, set same as vlan MTU
+        """
+        for iface in self._ifaces.values():
+            if (
+                iface.iface_type == InterfaceType.VLAN
+                or iface.iface_type == InterfaceType.VXLAN
+            ):
+
+                if iface.is_up:
+                    if iface.mtu:
```
Please merge all these conditions into single `if`

cathay4t commented on tests/lib/ifaces/vlan_iface_test.py:
```diff
@@ -20,7 +20,8 @@
 from libnmstate.error import NmstateValueError
-from libnmstate.schema import VLAN
+from libnmstate.ifaces import Ifaces
+from libnmstate.schema import VLAN, Interface
```
Current code style is import one thing per line. Please follow that.

cathay4t commented on tests/lib/ifaces/vxlan_iface_test.py:
```diff
@@ -20,7 +20,8 @@
 from libnmstate.error import NmstateValueError
-from libnmstate.schema import VXLAN
+from libnmstate.ifaces import Ifaces
+from libnmstate.schema import VXLAN, Interface
```
Current code style is import one thing per line. Please follow that.

adwait-thattey commented:
@cathay4t Done

adwait-thattey commented:
also rebased
